src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -0,0 +1,783 @@
+        /// <param name="hasNonzeroFractionalPart">whether there are nonzero digits after the decimal</param>
+        /// <param name="type">the kind of real number to build</param>
+        /// <param name="result">the result</param>
+        /// <returns>An indicator of the kind of result</returns>
+        private static SLD_STATUS ConvertBigIntegerToFloatingPointBits(byte[] integerValueAsBytes, uint integerBitsOfPrecision, bool hasNonzeroFractionalPart, FloatingPointType type, out ulong result)
+        {
+            int baseExponent = type.DenormalMantissaBits;
+            int exponent;
+            ulong mantissa;
+            bool has_zero_tail = !hasNonzeroFractionalPart;
+            int topElementIndex = ((int)integerBitsOfPrecision - 1) / 8;
+
+            // The high-order byte of integerValueAsBytes might not have a full eight bits.  However,  
+            // since the data are stored in quanta of 8 bits, and we really only need around 54  
+            // bits of mantissa for a double (and fewer for a float), we can just assemble data  
+            // from the eight high-order bytes and we will get between 59 and 64 bits, which is more  
+            // than enough.
+            int bottomElementIndex = Math.Max(0, topElementIndex - (64 / 8) + 1);
+            exponent = baseExponent + bottomElementIndex * 8;
+            mantissa = 0;
+            for (int i = (int)topElementIndex; i >= bottomElementIndex; i--)
+            {
+                mantissa <<= 8;
+                mantissa |= integerValueAsBytes[i];
+            }
+            for (int i = bottomElementIndex - 1; has_zero_tail && i >= 0; i--)
+            {
+                if (integerValueAsBytes[i] != 0) has_zero_tail = false;
+            }
+
+            return type.AssembleFloatingPointValue(mantissa, exponent, has_zero_tail, out result);
+        }
+
+        /// <summary>
+        /// Parse a sequence of digits into a BigInteger.
+        /// </summary>
+        /// <param name="data">The DecimalFloatingPointString containing the digits in its Mantissa</param>
+        /// <param name="integer_first_index">The index of the first digit to convert</param>
+        /// <param name="integer_last_index">The index just past the last digit to convert</param>
+        /// <returns>The BigInteger result</returns>
+        private static BigInteger AccumulateDecimalDigitsIntoBigInteger(DecimalFloatingPointString data, uint integer_first_index, uint integer_last_index)
+        {
+            if (integer_first_index == integer_last_index) return BigZero;
+            var valueString = data.Mantissa.Substring((int)integer_first_index, (int)(integer_last_index - integer_first_index));
+            return BigInteger.Parse(valueString);
+        }
+
+        /// <summary>
+        /// Return the number of significant bits set. 
+        /// </summary>
+        private static uint CountSignificantBits(ulong data)
+        {
+            uint result = 0;
```

Comment:
Isn't this popcnt?
Don't we have a helper for this (using ```data &= (data -1)``` within the loop body)?